utt/signature-processor/src/sigProcessor.cpp

```diff
@@ -0,0 +1,181 @@
+
+  uint64_t getSigId() const { return b()->sigId; }
+
+  void validate(const ReplicasInfo&) const override {}
+
+  static uint32_t sizeOfHeader() { return sizeof(Header); }
+
+ private:
+#pragma pack(push, 1)
+
+  struct Header : public bftEngine::impl::MessageBase::Header {
+    uint16_t genReplicaId;
+    uint64_t sigId;
+    uint16_t partialSigLength;
+  };
+#pragma pack(pop)
+  static_assert(sizeof(Header) == (6 + 2 + 8 + 2), "Header is 18B");
+
+  Header* b() const { return (Header*)msgBody_; }
+};
+}  // namespace messages
+SigProcessor::SigProcessor(uint16_t repId,
+                           uint16_t n,
+                           uint16_t threshold,
+                           uint64_t timer_handler_timeout,
+                           std::shared_ptr<bftEngine::impl::MsgsCommunicator> msgsCommunicator,
+                           std::shared_ptr<bftEngine::impl::MsgHandlersRegistrator> msgHandlers,
+                           concordUtil::Timers& timers)
+    : msgsCommunicator_{msgsCommunicator},
+      timers_{timers},
+      repId_{repId},
+      n_{n},
+      threshold_{threshold},
+      timer_handler_timeout_{timer_handler_timeout} {
+  msgHandlers->registerMsgHandler(bftEngine::impl::MsgCode::Reserved, [&](bftEngine::impl::MessageBase* message) {
+    messages::PartialSigMsg* msg = (messages::PartialSigMsg*)message;
+    onReceivingNewPartialSig(msg->getSigId(), msg->idOfGeneratedReplica(), msg->getPartialSig());
+    delete msg;
+  });
+  timeout_handler_ = timers_.add(std::chrono::milliseconds(timer_handler_timeout_),
+                                 concordUtil::Timers::Timer::RECURRING,
+                                 [&](concordUtil::Timers::Handle h) {
+                                   (void)h;
+                                   const auto entries = jobs_.getEntries();
+                                   for (const auto& [id, entry] : entries) {
+                                     (void)id;
+                                     if (entry.start_time == 0) continue;
+                                     auto now = std::chrono::steady_clock::now().time_since_epoch().count();
+                                     if (now - entry.start_time >= entry.job_timeout_ms_) {
+                                       onJobTimeout(entry);
+                                     }
+                                   }
+                                 });
+}
+void SigProcessor::processRCMSignature(uint64_t sigId,
+                                       const libutt::api::types::CurvePoint& pidHash,
+                                       const libutt::api::Commitment& rcm1,
+                                       const libutt::api::types::CurvePoint& s2,
+                                       std::shared_ptr<libutt::api::Registrator> signer,
+                                       uint64_t job_timeout) {
+  jobs_.registerEntry(sigId, [signer, rcm1, s2, pidHash](uint64_t rid, const libutt::api::types::Signature& sig) {
```

Comment:
why do you set a callback per signature and not have a map of callback per signature type that you set once e.g. sigTypeRCM -> validateRCM  etc

Reply:
the validation is per the specific request data, so if we don't want the sigCollector to know about the specific UTT type, we have to pass a callback (created by the caller)